README.md
[![ci-cd](https://github.com/UBC-MDS/bccovideda/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/bccovideda/actions/workflows/ci-cd.yml)
[![codecov](https://codecov.io/github/UBC-MDS/bccovideda/branch/main/graph/badge.svg)](https://codecov.io/github/UBC-MDS/bccovideda)
[![Documentation Status](https://readthedocs.org/projects/bccovideda/badge/?version=latest)](https://bccovideda.readthedocs.io/en/latest/?badge=latest)
# bccovideda

**Authors**:  Lianna Hovhannisyan, John Lee, Vadim Taskaev, Vanessa Yuen

The British Columbia Center for Disease Control (BCCDC) manages a range of provincial programs and clinics that contribute to public health and help control the spread of disease in BC. It administers and distributes the latest daily data on COVID-19 in British Columbia, which it provides in csv format along case-, lab- and regional-specific features as well as in comprehensive ArcGIS format via the [COVID-19 webpage](http://www.bccdc.ca/health-info/diseases-conditions/covid-19/data) (under "Download the data"). This package leverages daily case-specific COVID-19 data, allowing users to conveniently download the latest case data, and - per specified date range interval - compute several key statistics, visualize time series progression along age-related and regional parameters, and generate exploratory data analysis in the form of histogram figures supporting on-demand analysis. COVID-19 case detail parameters extracted using this package: 
- Reported_Date (in YYYY-MM-DD format)
- HA (provincial health region, e.g., "Vancouver Coast Health")
- Sex (M or F)
- Age_Group (reported along 10-yr age group bins, e.g., "60-69")
- Classification_Reported (diagnosis origin, e.g., "Lab-diagnosed")

## Installation

`bccovideda` can be installed from PyPI using the following terminal command:
```bash
$ pip install bccovideda
```

## Package Functions 

- `get_data()`
  - This function downloads the latest detailed daily case-specific COVID-19 from BCCDC's dedicated [COVID-19 homepage](http://www.bccdc.ca/health-info/diseases-conditions/covid-19/data). It returns a dataframe containing the extracted raw data. 

- `show_summary_stat()`
  - This function computes summary statistics from the available case-specific parameters, such as age-related and regional aggregate metrics. It returns a dataframe listing key identified summary statistics specified per the time interval queried. 

- `plot_line_by_date()`
  - This function returns a line chart plot of daily case counts, based on parameters and grouping selected by the user, per the time interval queried.

- `plot_hist_by_cond()`
  - This function returns a histogram plot based on parameters and grouping selected by the user, per the time interval queried, allowing for on-demand exploratory data analysis. 


## Usage

`bccovideda` can be used to download and compute summary statistics, generate exploratory data analysis histogram plots, and plot time series chart data as follows:
```python
from bccovideda.get_data import get_data
from bccovideda.show_summary_stat import show_summary_stat
from bccovideda.plot_hist_by_cond import plot_hist_by_cond
from bccovideda.plot_line_by_date import plot_line_by_date
```

```python
get_data()
```
<img src="https://github.com/UBC-MDS/bccovideda/raw/main/img/data.png" height="400">


```python
show_summary_stat("2022-01-01", "2022-01-13")
```
<img src="https://github.com/UBC-MDS/bccovideda/raw/main/img/summary.png" height="500">

```python
plot_hist_by_cond("2021-01-01", "2021-01-30", "Age")
```

!["Histogram"](https://github.com/UBC-MDS/bccovideda/raw/main/img/plot_histogram.png)


```python
plot_line_by_date("2021-01-01", "2021-01-30")
```
!["Line"](https://github.com/UBC-MDS/bccovideda/raw/main/img/plot_line.png)


## Role within Python Ecosystem

Given the relatively adequate accessibility of latest aggregate COVID-19 data combined with its persistent impact on socio-economics since early 2020, there are a number of rather comprehensive Python packages that perform similar data extract and exploratory data analysis functions, such as [covid](https://pypi.org/project/covid/), [covid19pyclient](https://pypi.org/project/covid19pyclient/), [covid19pandas](https://github.com/PayneLab/covid19pandas). In contrast to existing packages, `bccovideda` provides a simple user interface that  focuses on the localized provincial context of British Columbia, utilizing features specific to BCCDC's data administration conventions for generating a quick overview and on-demand analysis of trends and statistics pertaining to age-related and regional case characteristics.

## Dependencies

-   Python 3.9 and Python packages:

    -   pandas==1.3.5
    -   requests==2.27.1
    -   altair==4.2.0
    -   altair-saver==0.5.0

## Documentation

Documentation `bccovideda` can be found at [Read the Docs](https://bccovideda.readthedocs.io)

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## Contributors

Group 25 Contributors:
- Lianna Hovhannisyan: @liannah
- John Lee: @johnwslee
- Vadim Taskaev: @vtaskaev1
- Vanessa Yuen: @imtvwy

## License

The `bccovideda` project was created by DSCI 524 (Collaborative Software Development) Group 25 within the Master of Data Science program at the University of British Columbia (2021-2022). It is licensed under the terms of the MIT license.

## Credits

`bccovideda` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v1.1.0 (2022-02-03)
### Feature
* Get_data() exception handling ([`44092db`](https://github.com/UBC-MDS/bccovideda/commit/44092db701cfb705230254964a1a20699edea034))

## v1.0.10 (2022-01-29)


## v1.0.9 (2022-01-29)


## v1.0.8 (2022-01-29)


## v1.0.7 (2022-01-29)


## v1.0.6 (2022-01-28)


## v1.0.5 (2022-01-28)


## v1.0.4 (2022-01-28)


## v1.0.3 (2022-01-28)


## v1.0.2 (2022-01-28)


## v1.0.1 (2022-01-28)


## v0.1.0 (11/01/2022)

- First release of `bccovideda`!

test_bccovideda.py
from bccovideda import bccovideda


tests/test_get_data.py
#from bccovideda import bccovideda
from bccovideda.get_data import get_data

import pandas as pd
import os
from datetime import timedelta


def test_check_colnames():
    """Test that column header names of pd.DataFrame constructed 
    from downloaded csv file."""
    case_df_sample = pd.read_csv("tests/case_data_sample.csv")
    expected = case_df_sample.columns.values.tolist()
    actual = get_data().columns.values.tolist()
    assert actual == expected, "Error: Column headers of downloaded data do not match those of sample case data!"


def test_check_dtypes():
    """"Test that cases_df column dtypes are as expected."""
    case_df_sample = pd.read_csv(
        "tests/case_data_sample.csv", parse_dates=["Reported_Date"])
    expected = case_df_sample.dtypes
    actual = get_data().dtypes
    assert len(actual) == len(expected) and sorted(
        actual) == sorted(expected), "Error: Column dtypes of downloaded data do not match those of sample case data!"


def test_recent_data():
    """"Test that most recent data is not from 7 or more days ago."""
    expect = pd.to_datetime("now") - timedelta(days=7)
    actual = get_data()["Reported_Date"].max()
    assert actual > expect, "Error: Most recent data is from 7 or more days ago!"


tests/test_plot_hist_by_cond.py
import pytest
import pandas as pd
import altair as alt
import datetime
from datetime import date
from bccovideda.get_data import get_data
from bccovideda.plot_hist_by_cond import plot_hist_by_cond
alt.data_transformers.enable('data_server')
alt.renderers.enable('html')


def test_input_data():
    """
    Check Errors raised when incorrect inputs are used
    Test cases:
      1. input type
      2. input format
      3. input value

    """

    # test input type

    with pytest.raises(NameError):
        plot_hist_by_cond("2021-01-01", "2021-12-31", Region)

    with pytest.raises(TypeError):
        plot_hist_by_cond("2021-01-01", 2021-12-31, "Region")

    # test input format
    with pytest.raises(ValueError):
        plot_hist_by_cond("20210101", "20211231", "Age")

    with pytest.raises(ValueError):
        plot_hist_by_cond("2021/01/01", "2021/12/31", "Age")

    with pytest.raises(ValueError):
        plot_hist_by_cond("01-01-2021", "12-31-2021", "Age")

    with pytest.raises(ValueError):
        plot_hist_by_cond("Jan-01-2021", "Dec-31-2021", "Age")

    # test input value
    with pytest.raises(ValueError):
        plot_hist_by_cond("2021-13-01", "2021-12-31", "Age")

    with pytest.raises(ValueError):
        plot_hist_by_cond("2021-01-01", "2021-12-31", "age")

    with pytest.raises(ValueError):
        plot_hist_by_cond("2021-13-01", "2021-12-31", "region")

    with pytest.raises(ValueError):
        plot_hist_by_cond("2021-01-01", "2021-12-31", "area")


def test_plot_results():
    """
    Check if the returned plot is correct

    """
    plot_hist_age = plot_hist_by_cond('2021-10-28', '2021-12-28', "Age")
    plot_hist_region = plot_hist_by_cond('2021-10-28', '2021-12-28', "Region")
    assert str(type(plot_hist_age)) == "<class 'altair.vegalite.v4.api.Chart'>"
    assert str(type(plot_hist_region)
               ) == "<class 'altair.vegalite.v4.api.Chart'>"
    assert plot_hist_age.encoding.x.shorthand == 'count()', 'x_axis should be mapped to the x axis'
    assert plot_hist_age.encoding.y.shorthand == 'Age_Group', 'y_axis should be mapped to the y axis'
    assert plot_hist_region.encoding.x.shorthand == 'count()', 'x_axis should be mapped to the x axis'
    assert plot_hist_region.encoding.y.shorthand == 'HA', 'y_axis should be mapped to the y axis'
    assert plot_hist_age.mark == 'bar', 'mark should be a bar'
    assert plot_hist_region.mark == 'bar', 'mark should be a bar'


src/bccovideda/get_data.py

from operator import index
import requests
import os
import pandas as pd
import altair as alt
import datetime
alt.data_transformers.enable('data_server')


def get_data(url="http://www.bccdc.ca/Health-Info-Site/Documents/BCCDC_COVID19_Dashboard_Case_Details.csv", out_folder="data/raw"):
    """
    Downloads the entire "case details" data set from
    http://www.bccdc.ca/Health-Info-Site/Documents/BCCDC_COVID19_Dashboard_Case_Details.csv
    and returns it as a pandas data frame.

    Parameters
    ----------
    url        : string
                 URL to case data download page
    out_folder : string
                 Path to folder storing raw data set csv file

    Returns
    -------
    pandas.DataFrame : Dataframe containing entire BC Case Details Data

    Examples
    --------
    >>> get_data()
    """
    if not os.path.exists(os.getcwd() + "/" + out_folder):
        os.makedirs(out_folder + "/")

    if not isinstance(url, str):
        raise TypeError("url provided is not of string type")

    try:
        req = requests.get(url)
        if req.status_code == 200:
            pass
    except:
        raise ValueError("url provided is invalid")

    req = requests.get(url)

    url_content = req.content

    csv_file = open("data/raw/case_data.csv", "wb")
    csv_file.write(url_content)
    csv_file.close()

    cases_df = pd.read_csv("data/raw/case_data.csv",
                           parse_dates=['Reported_Date'])

    return cases_df


src/bccovideda/bccovideda.py

from operator import index
import requests
import os
import pandas as pd
import altair as alt
import datetime
alt.data_transformers.enable('data_server')


def get_data(url="http://www.bccdc.ca/Health-Info-Site/Documents/BCCDC_COVID19_Dashboard_Case_Details.csv", out_folder="data/raw"):
    """
    Downloads the entire "case details" data set from
    http://www.bccdc.ca/Health-Info-Site/Documents/BCCDC_COVID19_Dashboard_Case_Details.csv
    and returns it as a pandas data frame.

    Parameters
    ----------
    url        : string
                 URL to case data download page
    out_folder : string
                 Path to folder storing raw data set csv file

    Returns
    -------
    pandas.DataFrame : Dataframe containing entire BC Case Details Data

    Examples
    --------
    >>> get_data()
    """
    if not os.path.exists(os.getcwd() + "/" + out_folder):
        os.makedirs(out_folder + "/")

    req = requests.get(url)
    url_content = req.content

    csv_file = open("data/raw/case_data.csv", "wb")
    csv_file.write(url_content)
    csv_file.close()

    cases_df = pd.read_csv("data/raw/case_data.csv",
                           parse_dates=['Reported_Date'])

    cases_sample = pd.read_csv(
        "data/sample/case_data_sample.csv", parse_dates=['Reported_Date'])
    cases_sample.iloc[:100, 1:].to_csv(
        "data/sample/case_data_sample_noID.csv", index=False)

    return cases_df


def showSummaryStat(startDate, endDate):
    """
    Shows summary statistics for the Covid19 cases in BC for the period
    specified by startDate and endDate (format: YYYY-MM-DD).

    Parameters
    ----------
    startDate : string
                the start date of the period 
                (no earlier than '2020-01-29')
    endDate   : string
                the end date of the period 
                (no later than today)

    Returns
    -------
    pandas.DataFrame
        Data frame containing summary statistics with the followings:
        total_cases_count, latest_date, latest_daily_cases_count, 
        max_date, max_daily_cases_count, min_date, min_daily_cases_count, 
        max_age_group, max_age_group_count, min_age_group, min_age_group_count,
        max_region, max_region_count, min_region, min_region_count,

    Examples
    -------
    >>> showSummaryStat("2022-01-01", "2022-01-13")

    """


def plot_line_by_date(startDate, endDate, region='all'):
    """
    Plots the line chart of regional Covid19 cases over the period specified by 
    startDate and endDate (format: YYYY-MM-DD). The default argument value 
    for region is "all", showing the total number of Covid19 cases in BC.

    Parameters
    ----------
    startDate : string
                the start date of the period 
                (no earlier than '2020-01-29')
    endDate   : string
                the end date of the period 
                (no later than today)
    region    : list or str = 'all'
                Default value is string 'all' - displaying all regions.
                Other available values: combination of list of strings
                from available regions - Fraser, Vancouver Coastal, Vancouver Island, 
                Interior, Northern, Out of Canada

    Returns
    -------
    plot : altair.Chart object
           An altair plot object displaying line chart

    Examples
    --------
    >>> bccovideda.plotLineByDate("2021-01-01", "2021-12-31")
    >>> bccovideda.plotLineByDate("2021-01-01", "2021-12-31", region = ['Fraser'])
    """

    covid = get_data()

    # convert date column to string
    covid['Reported_Date'] = covid['Reported_Date'].apply(str)
    covid['Reported_Date'] = covid['Reported_Date'].str.slice(0, 10)

    # check the date format
    try:
        datetime.datetime.strptime(startDate, '%Y-%m-%d')
    except ValueError:
        raise ValueError("Incorrect date format, should be YYYY-MM-DD")
    try:
        datetime.datetime.strptime(endDate, '%Y-%m-%d')
    except ValueError:
        raise ValueError("Incorrect date format, should be YYYY-MM-DD")

    # check argument validity
    if not(isinstance(startDate, str)):
        raise TypeError('Invalid argument type: startDate must be a string.')
    elif not(isinstance(endDate, str)):
        raise TypeError('Invalid argument type: endDate must be a string.')
    elif not(isinstance(region, list) or region == 'all'):
        raise TypeError(
            'Invalid argument type: region must be a list or have a value `"all"`.')

    # check arguments value
    elif not(endDate <= covid.iloc[-1:, 0].values[0]):
        raise ValueError('Invalid argument value: endDate cannot be later '
                         'than the day the package is called.')
    elif not(startDate >= covid.iloc[0:, 0].values[0]):
        raise ValueError('Invalid argument value: startDate cannot be earlier '
                         'than the day the first case was recorded.')
    elif not(startDate < endDate):
        raise ValueError('Invalid argument value: endDate cannot be earlier '
                         'than the startDate.')
    elif not(set(region).issubset(set(pd.unique(covid['HA']))) or region == 'all'):
        raise ValueError('Invalid argument value: region must be valid BC region - '
                         'Either combination of `Fraser, Vancouver Coastal, Vancouver Island, '
                         'Interior, Northern, Out of Canada` or `all`')
    elif not(len(region) > 0):
        raise ValueError(
            'Invalid argument value: region cannot be an empty list')
    elif not(len(startDate) == 10 and len(endDate) == 10):
        raise ValueError('Invalid argument value: startDate and endDate format is '
                         '`YYYY-MM-DD` without spaces.'
                         )

    # filter the data
    if region == 'all':
        mask = ((covid["Reported_Date"] > startDate) &
                (covid["Reported_Date"] <= endDate))
    else:
        mask = ((covid["Reported_Date"] > startDate) &
                (covid["Reported_Date"] <= endDate) &
                covid["HA"].isin(region))

    # keep the filtered data
    temp = covid.loc[mask]

    # plot the line chart
    plot = (alt.Chart(temp,
                      title="Number of COVID19 cases over time")
            .mark_line().encode(
                x=alt.X("Reported_Date",
                        title="Date"),
                y=alt.Y("count()",
                        title="Number of Cases"),
                color=alt.Color("HA",
                                title='Region'))
            )

    return plot


def plotHistByCond(startDate, endDate, condition):
    """
    Plots the number of Covid19 cases with respect to the condition(per Age Group
    or Region) using histogram for the period specified by startDate and endDate

    Parameters
    ----------
    startDate : string
                the start date of the period 
                (no earlier than '2020-01-29')
    endDate   : string
                the end date of the period 
                (no later than today)
    condition : string
                Available condition - Age, Region

    Returns
    -------
    plot : altair.Chart object
           An altair plot object displaying histogram

    Examples
    --------
    >>> bccovideda.plotHistByCond("2021-01-01", "2021-12-31", "Age")
    """


src/bccovideda/plot_line_by_date.py

from bccovideda.get_data import get_data
import pandas as pd
import altair as alt
import datetime


def plot_line_by_date(startDate, endDate, region='all'):
    """
    Plots the line chart of regional Covid19 cases over the period specified by 
    startDate and endDate (format: YYYY-MM-DD). The default argument value 
    for region is "all", showing the total number of Covid19 cases in BC.

    Parameters
    ----------
    startDate : string
                the start date of the period 
                (no earlier than '2020-01-29')
    endDate   : string
                the end date of the period 
                (no later than today)
    region    : list or str = 'all'
                Default value is string 'all' - displaying all regions.
                Other available values: combination of list of strings
                from available regions - Fraser, Vancouver Coastal, Vancouver Island, 
                Interior, Northern, Out of Canada

    Returns
    -------
    plot : altair.Chart object
           An altair plot object displaying line chart

    Examples
    --------
    >>> plot_line_by_date("2021-01-01", "2021-12-31")
    >>> plot_line_by_date("2021-01-01", "2021-12-31", region = ['Fraser'])
    """

    covid = get_data()

    # convert date column to string
    covid['Reported_Date'] = covid['Reported_Date'].apply(str)
    covid['Reported_Date'] = covid['Reported_Date'].str.slice(0, 10)

    # check the date format
    try:
        datetime.datetime.strptime(startDate, '%Y-%m-%d')
    except ValueError:
        raise ValueError("Incorrect date format, should be YYYY-MM-DD")
    try:
        datetime.datetime.strptime(endDate, '%Y-%m-%d')
    except ValueError:
        raise ValueError("Incorrect date format, should be YYYY-MM-DD")

    # check argument validity
    if not(isinstance(startDate, str)):
        raise TypeError('Invalid argument type: startDate must be a string.')
    elif not(isinstance(endDate, str)):
        raise TypeError('Invalid argument type: endDate must be a string.')
    elif not(isinstance(region, list) or region == 'all'):
        raise TypeError(
            'Invalid argument type: region must be a list or have a value `"all"`.')

    # check arguments value
    elif not(endDate <= covid.iloc[-1:, 0].values[0]):
        raise ValueError('Invalid argument value: endDate cannot be later '
                         'than the day the package is called.')
    elif not(startDate >= covid.iloc[0:, 0].values[0]):
        raise ValueError('Invalid argument value: startDate cannot be earlier '
                         'than the day the first case was recorded.')
    elif not(startDate < endDate):
        raise ValueError('Invalid argument value: endDate cannot be earlier '
                         'than the startDate.')
    elif not(set(region).issubset(set(pd.unique(covid['HA']))) or region == 'all'):
        raise ValueError('Invalid argument value: region must be valid BC region - '
                         'Either combination of `Fraser, Vancouver Coastal, Vancouver Island, '
                         'Interior, Northern, Out of Canada` or `all`')
    elif not(len(region) > 0):
        raise ValueError(
            'Invalid argument value: region cannot be an empty list')
    elif not(len(startDate) == 10 and len(endDate) == 10):
        raise ValueError('Invalid argument value: startDate and endDate format is '
                         '`YYYY-MM-DD` without spaces.'
                         )

    # filter the data
    if region == 'all':
        mask = ((covid["Reported_Date"] > startDate) &
                (covid["Reported_Date"] <= endDate))
    else:
        mask = ((covid["Reported_Date"] > startDate) &
                (covid["Reported_Date"] <= endDate) &
                covid["HA"].isin(region))

    # keep the filtered data
    temp = covid.loc[mask]

    # plot the line chart
    plot = (alt.Chart(temp,
                      title="Number of COVID19 cases over time")
            .mark_line().encode(
                x=alt.X("Reported_Date",
                        title="Date"),
                y=alt.Y("count()",
                        title="Number of Cases"),
                color=alt.Color("HA",
                                title='Region'))
            )

    return plot


src/bccovideda/show_summary_stat.py
import pandas as pd
from datetime import date
from bccovideda.get_data import get_data


def show_summary_stat(startDate, endDate):
    """
    Shows summary statistics for the Covid19 cases in BC for the period
    specified by startDate and endDate (format: YYYY-MM-DD).

    Parameters
    ----------
    startDate : string
                the start date of the period 
                (no earlier than '2020-01-29')
    endDate   : string
                the end date of the period 
                (no later than today)


    Returns
    -------
    pandas.DataFrame
        Data frame containing summary statistics with the followings:
        total_cases_count, latest_date, latest_daily_cases_count, 
        max_date, max_daily_cases_count, min_date, min_daily_cases_count, 
        max_age_group, max_age_group_count, min_age_group, min_age_group_count,
        max_region, max_region_count, min_region, min_region_count
    None if there is no data matching the specified date range

    Examples
    -------
    >>> show_summary_stat("2022-01-01", "2022-01-13")

    """
    # input validation
    if not isinstance(startDate, str):
        raise TypeError("'startDate' should be of type 'str'.")
    if not isinstance(endDate, str):
        raise TypeError("'endDate' should be of type 'str'.")

    start_dt = date.fromisoformat(startDate)
    end_dt = date.fromisoformat(endDate)
    earliest_dt = date.fromisoformat("2020-01-29")

    if start_dt > end_dt:
        raise ValueError("startDate should not be later than endDate")
    if start_dt < earliest_dt:
        raise ValueError("startDate should not be earlier than 2020-01-29")
    if end_dt > date.today():
        raise ValueError("endDate should not be later than today")

    # download data
    df = get_data()

    mask = (df["Reported_Date"] >= startDate) & (
        df["Reported_Date"] <= endDate)
    df = df.loc[mask]

    # return None if it is an empty dataframe
    if len(df) == 0:
        return None

    # total_cases_count
    total = len(df)

    # count by date
    count_df = df["Reported_Date"].value_counts(sort=True, ascending=True)
    count_df_len = len(count_df)
    # latest_date, latest_daily_cases_count
    latest_date = count_df.index[count_df_len - 1]
    latest_date_count = count_df.iloc[count_df_len - 1]

    # max_date, max_daily_cases_count
    max_date = count_df.idxmax()
    max_daily_cases_count = count_df.max()
    # min_date, min_daily_cases_count
    min_date = count_df.idxmin()
    min_daily_cases_count = count_df.min()

    # count by age group
    age_df = df["Age_Group"].value_counts()
    # max_age_group, max_age_group_count
    max_age_group = age_df.idxmax()
    max_age_group_count = age_df.max()
    # min_age_group, min_age_group_count
    min_age_group = age_df.idxmin()
    min_age_group_count = age_df.min()

    # count by region
    region_df = df["HA"].value_counts()
    # max_region, max_region_count
    max_region = region_df.idxmax()
    max_region_count = region_df.max()

    # min_region, min_region_count,
    min_region = region_df.idxmin()
    min_region_count = region_df.min()

    summary_df = pd.DataFrame(
        {
            "total_cases_count": [total],
            "latest_date": [latest_date],
            "latest_daily_cases_count": [latest_date_count],
            "max_date": [max_date],
            "max_daily_cases_count": [max_daily_cases_count],
            "min_date": [min_date],
            "min_daily_cases_count": [min_daily_cases_count],
            "max_age_group": [max_age_group],
            "max_age_group_count": [max_age_group_count],
            "min_age_group": [min_age_group],
            "min_age_group_count": [min_age_group_count],
            "max_region": [max_region],
            "max_region_count": [max_region_count],
            "min_region": [min_region],
            "min_region_count": [min_region_count],
        }
    )

    return summary_df.T


src/bccovideda/plot_hist_by_cond.py
import pandas as pd
import altair as alt
import datetime
from datetime import date
from bccovideda.get_data import get_data


def plot_hist_by_cond(startDate, endDate, condition):
    """
    Plots the number of Covid19 cases with respect to the condition(per Age Group
    or Region) using histogram for the period specified by startDate and endDate

    Parameters
    ----------
    startDate : string
                the start date of the period 
                (no earlier than '2020-01-29')
    endDate   : string
                the end date of the period 
                (no later than today)
    condition : string
                Available condition - Age, Region

    Returns
    -------
    plot : altair.Chart object
           An altair plot object displaying histogram

    Examples
    --------
    >>> plot_hist_by_cond("2021-01-01", "2021-12-31", "Age")
    """

    # test input type
    if not isinstance(startDate, str):
        raise TypeError("'startDate' should be string.")
    if not isinstance(endDate, str):
        raise TypeError("'endDate' should be string.")
    if not isinstance(condition, str):
        raise TypeError("'endDate' should be string.")

    # test input format
    try:
        datetime.datetime.strptime(startDate, '%Y-%m-%d')
    except ValueError:
        raise ValueError("Incorrect date format, should be YYYY-MM-DD")
    try:
        datetime.datetime.strptime(endDate, '%Y-%m-%d')
    except ValueError:
        raise ValueError("Incorrect date format, should be YYYY-MM-DD")

    covid = get_data()

    start_dt = date.fromisoformat(startDate)
    end_dt = date.fromisoformat(endDate)
    earliest_dt = date.fromisoformat("2020-01-29")

    # test input value
    if start_dt > end_dt:
        raise ValueError("startDate should not be later than endDate")
    if start_dt < earliest_dt:
        raise ValueError("startDate should not be earlier than 2020-01-29")
    if end_dt > date.today():
        raise ValueError("endDate should not be later than today")

    mask = (covid["Reported_Date"] > startDate) & (
        covid["Reported_Date"] <= endDate)
    temp = covid.loc[mask]

    age_group_label = [
        '90+', '80-89', '70-79', '60-69', '50-59', '40-49', '30-39',
        '20-29', '10-19', '<10'
    ]

    region_label = [
        'Fraser', 'Vancouver Coastal', 'Vancouver Island', 'Interior',
        'Northern', 'Out of Canada'
    ]

    plot = 0   # plot initialized to prevent UnboundLocalError

    if condition == "Age":
        plot = alt.Chart(temp, title="Number of Cases by Age Group").mark_bar().encode(
            y=alt.Y("Age_Group", sort=age_group_label, title="Age Group"),
            x=alt.X("count()", title="Number of Cases"))
    elif condition == "Region":
        plot = alt.Chart(temp, title="Number of Cases by Region").mark_bar().encode(
            y=alt.Y("HA", sort=region_label, title="Region"),
            x=alt.X("count()", title="Number of Cases"))
    else:
        raise ValueError('Condition should be "Age" or "Region"')

    return plot
